// @vite-plugin-vercel/vike/dist/vike.js
// vike.ts
import { prerender as prerenderCli } from "vike/prerender";
import fs from "fs/promises";
import path from "path";
import { fileURLToPath } from "url";
import { normalizePath } from "vite";
import "vike/__internal/setup";
import {
  getPagesAndRoutes,
  route
} from "vike/__internal";
import { nanoid } from "nanoid";

// route-regex.ts
function getSegmentRegex(segment) {
  if (segment.startsWith("@")) {
    return "/[^/]+";
  } else if (segment === "*") {
    return "/.+?";
  }
  return `/` + segment;
}
function getParametrizedRoute(route2) {
  const segments = (route2.replace(/\/$/, "") || "/").slice(1).split("/");
  return segments.map(getSegmentRegex).join("");
}

// vike.ts
import { newError } from "@brillout/libassert";
var libName = "vite-plugin-vercel:vike";
var rendererDestination = "ssr_";
var __dirname = path.dirname(fileURLToPath(import.meta.url));
function assert(condition, errorMessage) {
  if (condition) {
    return;
  }
  const err = newError(`[${libName}][Wrong Usage] ${errorMessage}`, 2);
  throw err;
}
function getRoot(config) {
  return normalizePath(config.root || process.cwd());
}
function getOutDirRoot(config) {
  const outDir = config.build.outDir;
  return outDir.endsWith("/server") || outDir.endsWith("/client") ? path.normalize(path.join(outDir, "..")) : outDir;
}
function getOutput(config, suffix) {
  var _a, _b;
  return path.join(
    ((_a = config.vercel) == null ? void 0 : _a.outDir) ? "" : getRoot(config),
    ((_b = config.vercel) == null ? void 0 : _b.outDir) ?? ".vercel/output",
    suffix ?? ""
  );
}
function getOutDir(config, force) {
  const p = path.join(config.root, normalizePath(config.build.outDir));
  if (!force)
    return p;
  return path.join(path.dirname(p), force);
}
async function copyDir(src, dest) {
  await fs.mkdir(dest, { recursive: true });
  const entries = await fs.readdir(src, { withFileTypes: true });
  for (const entry of entries) {
    const srcPath = path.join(src, entry.name);
    const destPath = path.join(dest, entry.name);
    if (entry.isDirectory()) {
      await copyDir(srcPath, destPath);
    } else {
      await fs.copyFile(srcPath, destPath);
    }
  }
}
function assertIsr(resolvedConfig, exports) {
  var _a, _b, _c;
  if (exports === null || typeof exports !== "object")
    return null;
  if (!("isr" in exports))
    return null;
  const isr = exports.isr;
  assert(
    typeof isr === "boolean" || typeof isr === "object" && typeof isr.expiration === "number" && isr.expiration > 0,
    ` \`{ expiration }\` must be a positive number`
  );
  if (isr === true) {
    assert(
      typeof ((_a = resolvedConfig.vercel) == null ? void 0 : _a.expiration) === "number" && ((_b = resolvedConfig.vercel) == null ? void 0 : _b.expiration) > 0,
      "`export const isr = true;` requires a default positive value for `expiration` in vite config"
    );
    return (_c = resolvedConfig.vercel) == null ? void 0 : _c.expiration;
  }
  return isr.expiration;
}
var prerender = async (resolvedConfig) => {
  const routes = {};
  await prerenderCli({
    viteConfig: {
      root: getRoot(resolvedConfig),
      build: {
        outDir: getOutDirRoot(resolvedConfig)
      }
    },
    async onPagePrerender(pageContext) {
      var _a;
      const { filePath, fileContent } = pageContext._prerenderResult;
      const isr = assertIsr(resolvedConfig, pageContext.exports);
      delete pageContext.urlPathname;
      delete pageContext.urlParsed;
      const foundRoute = await route(pageContext);
      if (!pageContext.is404) {
        assert(foundRoute, `Page with id ${pageContext._pageId} not found`);
        const routeMatch = (_a = foundRoute.pageContextAddendum._routeMatches) == null ? void 0 : _a[0];
        if (typeof isr === "number" && routeMatch && typeof routeMatch !== "string" && routeMatch.routeType === "FILESYSTEM") {
          return;
        }
      }
      const relPath = path.relative(
        getOutDir(resolvedConfig, "client"),
        filePath
      );
      const newFilePath = path.join(
        getOutput(resolvedConfig, "static"),
        relPath
      );
      const parsed = path.parse(relPath);
      const pathJoined = parsed.name === "index" ? parsed.dir : path.join(parsed.dir, parsed.name);
      if (relPath.endsWith(".html")) {
        routes[relPath] = {
          path: pathJoined === "index" ? "" : pathJoined
        };
      }
      await fs.mkdir(path.dirname(newFilePath), { recursive: true });
      await fs.writeFile(newFilePath, fileContent);
    }
  });
  return routes;
};
function getRouteDynamicRoute(pageRoutes, pageId) {
  for (const route2 of pageRoutes) {
    if (route2.pageId === pageId) {
      if (route2.routeType === "STRING") {
        return getParametrizedRoute(route2.routeString);
      } else if (route2.routeType === "FUNCTION") {
        return () => {
        };
      }
    }
  }
  return null;
}
function getRouteFsRoute(pageRoutes, pageId) {
  for (const route2 of pageRoutes) {
    if (route2.pageId === pageId && route2.routeType === "FILESYSTEM") {
      return route2.routeString;
    }
  }
  return null;
}
async function getSsrEndpoint(userConfig, source) {
  const sourcefile = source ?? path.join(__dirname, "..", "templates", "ssr_.template.ts");
  const contents = await fs.readFile(sourcefile, "utf-8");
  const resolveDir = path.dirname(sourcefile);
  return {
    source: {
      contents,
      sourcefile,
      loader: sourcefile.endsWith(".ts") ? "ts" : sourcefile.endsWith(".tsx") ? "tsx" : sourcefile.endsWith(".js") ? "js" : sourcefile.endsWith(".jsx") ? "jsx" : "default",
      resolveDir
    },
    destination: rendererDestination,
    addRoute: false
  };
}
function vitePluginSsrVercelPlugin(options = {}) {
  return {
    name: libName,
    apply: "build",
    async config(userConfig) {
      var _a, _b, _c, _d, _e;
      if (!((_a = userConfig.build) == null ? void 0 : _a.ssr))
        return {};
      const additionalEndpoints = ((_c = (_b = userConfig.vercel) == null ? void 0 : _b.additionalEndpoints) == null ? void 0 : _c.flatMap((e) => e.destination).some((d) => d === rendererDestination)) ? (_d = userConfig.vercel) == null ? void 0 : _d.additionalEndpoints : [await getSsrEndpoint(userConfig)];
      return {
        vitePluginSsr: {
          prerender: {
            disableAutoRun: true
          }
        },
        vercel: {
          prerender: ((_e = userConfig.vercel) == null ? void 0 : _e.prerender) ?? prerender,
          additionalEndpoints,
          rewrites: [
            {
              source: options.source ? `(${options.source})` : "((?!/api).*)",
              destination: `/${rendererDestination}/?__original_path=$1`,
              enforce: "post"
            }
          ]
        }
      };
    }
  };
}
function findPageFile(pageId, pageFilesAll) {
  return pageFilesAll.find(
    (p) => p.pageId === pageId && p.fileType === ".page"
  );
}
function vitePluginVercelVpsIsrPlugin() {
  return {
    name: "vite-plugin-vercel:vps-isr",
    apply: "build",
    async config(userConfig) {
      async function getPagesWithConfigs() {
        const { pageFilesAll, allPageIds, pageRoutes, pageConfigs } = await getPagesAndRoutes();
        const isLegacy = pageFilesAll.length > 0;
        if (isLegacy) {
          await Promise.all(pageFilesAll.map((p) => {
            var _a;
            return (_a = p.loadFile) == null ? void 0 : _a.call(p);
          }));
        }
        return await Promise.all(
          allPageIds.map(async (pageId) => {
            let page;
            if (isLegacy) {
              const _page = await findPageFile(pageId, pageFilesAll);
              assert(
                _page,
                `Cannot find page ${pageId}. Contact the vite-plugin-vercel maintainer on GitHub / Discord`
              );
              page = {
                config: _page.fileExports,
                filePath: _page.filePath
              };
            } else {
              const pageConfig = pageConfigs.find((p) => p.pageId === pageId);
              assert(
                pageConfig,
                `Cannot find page config ${pageId}. Contact the vite-plugin-vercel maintainer on GitHub / Discord`
              );
              const simplePageConfig = {};
              for (const [k, v] of Object.entries(pageConfig.configValues)) {
                simplePageConfig[k] = v.value;
              }
              page = {
                config: simplePageConfig,
                filePath: pageConfig.pageId
              };
            }
            const route2 = getRouteDynamicRoute(pageRoutes, pageId) ?? getRouteFsRoute(pageRoutes, pageId);
            let isr = assertIsr(userConfig, page.config);
            if (typeof route2 === "function" && isr) {
              console.warn(
                `Page ${pageId}: ISR is not supported when using route function. Remove \`{ isr }\` export or use a route string if possible.`
              );
              isr = null;
            }
            return {
              _pageId: pageId,
              // used for debug purpose
              filePath: page.filePath,
              isr,
              route: typeof route2 === "string" ? getParametrizedRoute(route2) : null
            };
          })
        );
      }
      return {
        vercel: {
          isr: async () => {
            var _a;
            let userIsr = {};
            if ((_a = userConfig.vercel) == null ? void 0 : _a.isr) {
              if (typeof userConfig.vercel.isr === "function") {
                userIsr = await userConfig.vercel.isr();
              } else {
                userIsr = userConfig.vercel.isr;
              }
            }
            const pagesWithConfigs = await getPagesWithConfigs();
            return pagesWithConfigs.filter((p) => typeof p.isr === "number").reduce((acc, cur) => {
              const path2 = cur._pageId.replace(/\/index$/g, "") + "-" + nanoid();
              acc[path2] = {
                expiration: cur.isr,
                symlink: rendererDestination,
                route: cur.route ? cur.route + "(?:\\/index\\.pageContext\\.json)?" : void 0
              };
              return acc;
            }, userIsr);
          }
        }
      };
    }
  };
}
function vitePluginVercelVpsCopyStaticAssetsPlugins() {
  let resolvedConfig;
  return {
    apply: "build",
    name: "vite-plugin-vercel:vps-copy-static-assets",
    enforce: "post",
    configResolved(config) {
      resolvedConfig = config;
    },
    async closeBundle() {
      var _a;
      if (!((_a = resolvedConfig.build) == null ? void 0 : _a.ssr))
        return;
      await copyDistClientToOutputStatic(resolvedConfig);
    }
  };
}
async function copyDistClientToOutputStatic(resolvedConfig) {
  await copyDir(
    getOutDir(resolvedConfig, "client"),
    getOutput(resolvedConfig, "static")
  );
}
function allPlugins(options = {}) {
  return [
    vitePluginVercelVpsIsrPlugin(),
    vitePluginSsrVercelPlugin(options),
    vitePluginVercelVpsCopyStaticAssetsPlugins()
  ];
}
export {
  assert,
  allPlugins as default,
  getOutDir,
  getOutput,
  getRoot,
  getSsrEndpoint,
  prerender,
  vitePluginSsrVercelPlugin,
  vitePluginVercelVpsCopyStaticAssetsPlugins,
  vitePluginVercelVpsIsrPlugin
};
