// vike/dist/cjs/shared/page-configs/assertExports.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertExportsOfConfigFile = exports.assertExportsOfValueFile = void 0;
const utils_js_1 = require("../utils.js");
const picocolors_1 = __importDefault(require("@brillout/picocolors"));
const EXPORTS_IGNORE = [
    // vite-plugin-solid adds `export { $$registrations }`
    '$$registrations',
    // @vitejs/plugin-vue adds `export { _rerender_only }`
    '_rerender_only'
];
// Tolerate `export { frontmatter }` in .mdx files
const TOLERATE_SIDE_EXPORTS = ['.md', '.mdx'];
function assertExportsOfValueFile(fileExports, filePathToShowToUser, configName) {
    assertExports(fileExports, filePathToShowToUser, configName);
}
exports.assertExportsOfValueFile = assertExportsOfValueFile;
function assertExportsOfConfigFile(fileExports, filePathToShowToUser) {
    assertExports(fileExports, filePathToShowToUser);
    const exportDefault = fileExports.default;
    (0, utils_js_1.assertUsage)((0, utils_js_1.isObject)(exportDefault), `The ${picocolors_1.default.cyan('export default')} of ${filePathToShowToUser} should be an object (but it's ${picocolors_1.default.cyan(`typeof exportDefault === ${JSON.stringify(typeof exportDefault)}`)} instead)`);
}
exports.assertExportsOfConfigFile = assertExportsOfConfigFile;
function assertExports(fileExports, filePathToShowToUser, configName) {
    const exportsAll = Object.keys(fileExports);
    const exportsRelevant = exportsAll.filter((exportName) => !EXPORTS_IGNORE.includes(exportName));
    const exportsInvalid = exportsRelevant.filter((e) => e !== 'default' &&
        // !!configName => isValueFile
        e !== configName);
    if (exportsInvalid.length === 0) {
        if (exportsRelevant.length === 1) {
            return;
        }
        const exportDefault = picocolors_1.default.cyan('export default');
        const exportConfigName = picocolors_1.default.cyan(`export { ${configName} }`);
        if (exportsRelevant.length === 0) {
            let errMsg = `${filePathToShowToUser} doesn't export any value, but it should have a ${exportDefault}`;
            if (configName)
                errMsg += ` or ${exportConfigName}`;
            (0, utils_js_1.assertUsage)(false, errMsg);
        }
        else {
            (0, utils_js_1.assert)(exportsRelevant.length === 2);
            (0, utils_js_1.assertWarning)(false, `${filePathToShowToUser} remove ${exportConfigName} or ${exportDefault}`, {
                onlyOnce: true
            });
        }
    }
    else {
        // !configName => isConfigFile
        if (!configName) {
            const exportsInvalidStr = exportsInvalid.join(', ');
            (0, utils_js_1.assertUsage)(false, `${filePathToShowToUser} replace ${picocolors_1.default.cyan(`export { ${exportsInvalidStr} }`)} with ${picocolors_1.default.cyan(`export default { ${exportsInvalidStr} }`)}`);
        }
        // !!configName => isValueFile
        else {
            if (TOLERATE_SIDE_EXPORTS.some((ext) => filePathToShowToUser.endsWith(ext)))
                return;
            exportsInvalid.forEach((exportInvalid) => {
                (0, utils_js_1.assertWarning)(false, `${filePathToShowToUser} should have only a single export: move ${picocolors_1.default.cyan(`export { ${exportInvalid} }`)} to +config.h.js or its own +${exportsInvalid}.js`, { onlyOnce: true });
            });
        }
    }
}
