// vike/dist/cjs/shared/page-configs/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfigValue = exports.getDefinedAtString = exports.getConfigDefinedAtString = exports.getHookFilePathToShowToUser = exports.getConfigValueFilePathToShowToUser = exports.getPageConfig = void 0;
var getConfigDefinedAtString_js_1 = require("./helpers/getConfigDefinedAtString.js");
Object.defineProperty(exports, "getConfigDefinedAtString", { enumerable: true, get: function () { return getConfigDefinedAtString_js_1.getConfigDefinedAtString; } });
Object.defineProperty(exports, "getDefinedAtString", { enumerable: true, get: function () { return getConfigDefinedAtString_js_1.getDefinedAtString; } });
var getConfigValue_js_1 = require("./helpers/getConfigValue.js");
Object.defineProperty(exports, "getConfigValue", { enumerable: true, get: function () { return getConfigValue_js_1.getConfigValue; } });
const utils_js_1 = require("../utils.js");
function getPageConfig(pageId, pageConfigs) {
    const pageConfig = pageConfigs.find((p) => p.pageId === pageId);
    (0, utils_js_1.assert)(pageConfigs.length > 0);
    (0, utils_js_1.assert)(pageConfig);
    return pageConfig;
}
exports.getPageConfig = getPageConfig;
function getConfigValueFilePathToShowToUser({ definedAt }) {
    // A unique file path only exists if the config value isn't cumulative nor computed:
    //  - cumulative config values have multiple file paths
    //  - computed values don't have any file path
    if ('isComputed' in definedAt || 'files' in definedAt)
        return null;
    const { filePathToShowToUser } = definedAt;
    (0, utils_js_1.assert)(filePathToShowToUser);
    return filePathToShowToUser;
}
exports.getConfigValueFilePathToShowToUser = getConfigValueFilePathToShowToUser;
function getHookFilePathToShowToUser({ definedAt }) {
    const filePathToShowToUser = getConfigValueFilePathToShowToUser({ definedAt });
    (0, utils_js_1.assert)(filePathToShowToUser);
    return filePathToShowToUser;
}
exports.getHookFilePathToShowToUser = getHookFilePathToShowToUser;
